### ISSUE TITLE
Add scoped variables as an alternative to nested Using

### ISSUE THREAD

**Issue by jdmichel** [dotnet/vblang#387]
This is another recent C# feature that I've always hoped to bring from C++.
(See [using declarations](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#using-declarations))
```
Using a = F1()
Using b = F2()
Using c = F3()
...
```
Instead of:
```
Using a = F1()
    Using b = F2()
          Using c = F3()
          …
          End Using
    End Using
End Using
```

The idea is that a, b, and c are Disposed when the variables go out of scope. It avoids the need to nest Using statements. 

I can't see any way to add this enhancement to the existing Using keyword in VB, but maybe a different keyword would be even more clear.

Scoped a = F1()
Scoped b = F2()
Scoped c = F3()

**Comment by zspitz** [comment]
@jdmichel 

> I can't see any way to add this enhancement to the existing Using keyword in VB

Why not?
```vb
Using a = F1
Using b = F2
Using c = F3
' more stuff here
```

**Comment by reduckted** [comment]
Interesting. Can't say I'm a fan of it, but I can see why some people might want it.

For what it's worth, this:
```vb
Using a = F1()
    Using b = F2()
          Using c = F3()
            ' ...
          End Using
    End Using
End Using
```

Can already be simplified to this:

```vb
Using a = F1(), b = F2(), c = F3()
    ' ...
End Using
```

It produces exactly the same IL. It would also work if you had to pass `a` and `b` as parameters to the other functions:

```vb
' This:
Using a = F1()
    Using b = F2(a)
          Using c = F3(b)
            ' ...
          End Using
    End Using
End Using

' Is exactly the same as this:
Using a = F1(), b = F2(a), c = F3(b)
    ' ...
End Using
```

> @zspitz Why not?
> ```vb
> Using a = F1
> ```

Because it would be a breaking change. Even if the compiler was able to understand that there is a missing `End Using` statement and handle it appropriately, it would still break the tooling. Today, when you type `Using a = F1()`  and press enter, the `End Using` statement is automatically added. Either the tooling would continues to add the `End Using` statement, and then you have to manually delete it, or the tooling stops adding the `End Using` statement and it breaks existing behavior, which will not go down well.

Why not just follow the C# lead and put `Using` before the `Dim` keyword?

```vb
Using Dim a = F1()
Using Dim b = F2()
Using Dim c = F3()
```

**Comment by zspitz** [comment]
@reduckted 

> Even if the compiler was able to understand that there is a missing End Using statement and handle it appropriately, it would still break the tooling

Agreed. However, there's already precedent for something similar. `If` statements have two forms -- the single line form, and the multi-line form. When the programmer types `If ... Then`, nothing happens until the user presses ENTER.

If we have some sort of keyword at the end of the first line of the single-line `Using` (like the `Then` of `If. ... Then`), then when the user presses ENTER without that keyword, the IDE would fill in the `End Using`:
```vb
Using a = F1() Do
Using a = F2() Do
Using a = F3() Do
```
vs
```vb
Using a = F1
End Using
```

> Why not just follow the C# lead and put Using before the Dim keyword?

Because in C# both forms of `using` have a variable clause, not just the proposed scoped version. I think introducing it into VB.NET only in the case of a scoped-variable `Using` would be too confusing.

---

Another alternative might be something along [the way it's done in F#](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/resource-management-the-use-keyword). There's the block form:
```
using (System.IO.File.CreateText(filename)) ( fun file1 ->
    file1.WriteLine("{0}", obj.ToString() )
)
```
and the scoped-variable form:
```
use file1 = File.CreateText(filename)
file1.WriteLine("{0}", obj.ToString() )
```
Something similar in VB.NET:
```vb
Use a = F1()
Use b = F2
```

**Comment by pricerc** [comment]
Apart from an `End Using`, what does this save ? 

I'm not a fan, that C# example has a code smell about it that I can't quite figure out. 

I guess I figure that it's bad form to not clean up after yourself. If you open a file, you should jolly well close if after you and not expect someone else to do it for you. And if you were writing to the file, how does the compiler know whether to save or discard the file?

**Comment by zspitz** [comment]
@pricerc 

> I guess I figure that it's bad form to not clean up after yourself .... and not expect someone else to do it for you.

And if I've created a `List(Of String)`, make sure to set it to `Nothing` when you're done with it; it's pure laziness to depend on the garbage collector to clean up after me. :)

But the point is the file will be closed once the variable goes out of scope of the function or block. Admittedly, if the rest of the function may take some time, it's better to use the block form, and explicitly close the file. But otherwise, why not?

> what does this save?

It saves an extra level of indentation.

**Comment by jdmichel** [comment]
I really like "Use". Why didn't I think of that? I originally posted this feature on my blog before VB had the Using keyword, and I guess I was stuck in my thinking. Thanks.

**Comment by pricerc** [comment]
I remain firmly on the fence. 

Although I can understand the C# peeps' desire to lose curly brackets - they can be a right pain to keep track of, I think there is some loss of visibility of the semantics of `using`. 

In languages without well-defined `End` clauses, you don't lose much information by discarding them, but in VB, the `End` family are important aids to legibility and future maintenance.

The `Using` and `End Using` block has a special purpose, it's not just a variation on declaring a variable. It's presence tells you: "Arrr, here be dragons!", that there are unmanaged resources around, that you shouldn't hold any longer than necessary. And `End Using` tells the compiler at what point they should be gotten rid of. 

And it's also a hand-dandy alternative to the previouslyrequiredtoachievethesameresult(tm) try/finally construct (which you still really need to use if you need to `catch` something). For those who came late, a refresher from the VB language reference [using statement](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/statements/using-statement) :
```VB
Using resource As New resourceType   
    ' Insert code to work with resource.  
End Using  
  
' For the acquisition and disposal of resource, the following  
' Try construction is equivalent to the Using block.  
Dim resource As New resourceType  
Try   
    ' Insert code to work with resource.  
Finally   
    If resource IsNot Nothing Then  
        resource.Dispose()   
    End If  
End Try   
```

**Comment by zspitz** [comment]
> And it's also a hand-dandy alternative to the previouslyrequiredtoachievethesameresult(tm) try/finally construct

A single line `Using` syntax -- e.g. `Use` -- would still provide the same benefit. This:
```vb
Sub Foo()
    Use conn = New SqlConnection
    ' do stuff
End Sub
```
would be equivalent to:
```vb
Sub Foo()
    Using conn  = New SqlConnection
        ' do stuff
    End Using
End Sub
```

**Comment by pricerc** [comment]
@zspitz 

> 
> A single line `Using` syntax -- e.g. `Use` -- would still provide the same benefit. This:
> 
> ```vbnet
> Sub Foo()
>     Use conn = New SqlConnection
>     ' do stuff
> End Sub
> ```
> 

I get that. But then why not just have the compiler detect the presence of IDisposable on a variable and do it automatically?

**Comment by jdmichel** [comment]
> I get that. But then why not just have the compiler detect the presence of IDisposable on a variable and do it automatically?

I thought about that too actually. That might be worth considering, but it's a much bigger change. One problem is that sometimes people make things Disposable which should not be disposed. (e.g. HttpClient).  But maybe it's worth it for VB to just make things simpler, and always call a Dispose method on a reference object when it goes out of scope? Not sure if that would really be safe. I'm afraid there could be times when you wouldn't want it, but I can't think of any at the moment. 

I guess I've just run into more situations on projects where I found myself with deeply nested using statements to handle a bunch of things that needed disposing. I could probably have just added them as comma separated statements, but "Use" seems more line-oriented which is one of the main things I like about VB. I dislike features that encourage long run-on "sentences" in source code.

**Comment by pricerc** [comment]
Why should HttpClient not be disposed if it's going out of scope? 

If you want to use it again later, can't you just widen the scope?

**Comment by zspitz** [comment]
@pricerc @jdmichel 

> But then why not just have the compiler detect the presence of `IDisposable` on a variable and do it automatically?

I'm just guessing, but mightn't the object behind the variable have other references? In that case `Dispose` shouldn't be called even if the object goes out of scope. And detecting whether a given object has any active references is something that requires the GC, so we're back to the basic problem which `Using` is trying to solve -- resources are kept open until the GC comes around.

A refinement might be that any `IDisposble` object created within this method should automatically be `Disposed` when the local variables that reference it go out of scope at the end of the method / block. But that means detecting that it hasn't left the scope of the method in other ways -- not returned, not set to a field / locat variable in an outer scope, set to a `ByRef` variable.

Moreover, objects can be created without an explicit constructor -- when returned be another method. How could we detect object creation in another method?
```vb
Dim conn = CreateMySqlConnection
```

**Comment by reduckted** [comment]
> @jdmichel ...and always call a Dispose method on a reference object when it goes out of scope?

This is a terrible idea. An example is the easiest way to explain:

```vb
Public Function CreateStuff() As Stuff
    Dim thingThatCanBeDisposed As IDisposable
    Dim stuff As Stuff

    thingThatCanBeDisposed = New DisposableThing()
    stuff = New Stuff(thingThatCanBeDisposed)

    Return stuff
    ' `thingThatCanBeDisposed` is now out of scope. 
    ' Let's dispose it and break `Stuff` while we're at it.
End Sub
```

**Comment by jdmichel** [comment]
> Why should HttpClient not be disposed if it's going out of scope?
> If you want to use it again later, can't you just widen the scope?

[Why HttpClient should not be disposed](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/)

**Comment by jdmichel** [comment]
> This is a terrible idea. An example is the easiest way to explain:

I think I agree. I just couldn't think of an example. Even in your example, it seems like the same sort of logic the garbage detector uses, could tell that `thingThatCanBeDisposed` and `stuff` both "escape" from the local scope and therefore can't be disposed. 

I still prefer things like this to be explicit, and think `Use` could help with the problem of deeply nested `Using` blocks.

**Comment by pricerc** [comment]
> 
> > Why should HttpClient not be disposed if it's going out of scope?
> > If you want to use it again later, can't you just widen the scope?
> 
> [Why HttpClient should not be disposed](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/)

Actually, that article supports my assertion that you should widen the scope - declare your HttpClient as a *static*(shared in VB), class-level field.

**Comment by pricerc** [comment]
> 
> 
> > @jdmichel ...and always call a Dispose method on a reference object when it goes out of scope?
> 
> This is a terrible idea. An example is the easiest way to explain:
> 
> ```vbnet
> Public Function CreateStuff() As Stuff
>     Dim thingThatCanBeDisposed As IDisposable
>     Dim stuff As Stuff
> 
>     thingThatCanBeDisposed = New DisposableThing()
>     stuff = New Stuff(thingThatCanBeDisposed)
> 
>     Return stuff
>     ' `thingThatCanBeDisposed` is now out of scope. 
>     ' Let's dispose it and break `Stuff` while we're at it.
> End Sub
> ```

Ok, but if thingThatCanBeDisposed was changed to a Non-IDisposable object, it wouldn't get destroyed when the method exits (i.e. the caller of CreateStuff will still be able to use it), because the GC knows that it is still in use. Why could a similar process not be used for IDisposable objects?

**Comment by reduckted** [comment]
> @jdmichel  [Why HttpClient should not be disposed](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/)

> @pricerc  Actually, that article supports my assertion that you should widen the scope - declare your HttpClient as a _static_(shared in VB), class-level field.

Yes, that article is a bit misleading. It's stating that you should use _one_ instance of `HttpClient` everywhere. Because you use that _one_ instance everywhere, you don't call `Dispose()` once you've finished using it, because something else will need to use it later.

**Comment by reduckted** [comment]
> @pricerc Ok, but if thingThatCanBeDisposed was changed to a Non-IDisposable object, it wouldn't get destroyed when the method exits (i.e. the caller of CreateStuff will still be able to use it), because the GC knows that it is still in use.

You're confusing IDisposable with garbage collection. In that example, it doesn't matter whether `thingThatCanBeDisposed` is `IDisposable` or not, the garbage collector is _not_ going to collect it when `CreateStuff()` returns because a reference to it still exists inside the `Stuff` instance.

Here's a detailed Stack Overflow answer about `IDisposable`: https://stackoverflow.com/a/538238/4397397

> @pricerc Why could a similar process not be used for IDisposable objects?

If a similar process could exist, then it would, and there would be no need for the `IDisposable` interface. What you're asking for here is a change to the runtime, _not_ to the VB language.

**Comment by pricerc** [comment]
> 
> 
> > @jdmichel  [Why HttpClient should not be disposed](https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/)
> 
> > @pricerc  Actually, that article supports my assertion that you should widen the scope - declare your HttpClient as a _static_(shared in VB), class-level field.
> 
> Yes, that article is a bit misleading. It's stating that you should use _one_ instance of `HttpClient` everywhere. Because you use that _one_ instance everywhere, you don't call `Dispose()` once you've finished using it, because something else will need to use it later.

I wouldn't say it's misleading at all. It was just demonstrating a high-performance technique when using HttpClient, that also supports the argument that IDisposable resources that have use outside of the scope in which they're instantiated should be considered candidates for scope-widening.

Is there some other reason why you wouldn't Dispose an HttpClient instance when returning from a method? (other than if you're returning that instance from the method)

**Comment by pricerc** [comment]
Firstly:

>  What you're asking for here is a change to the runtime, _not_ to the VB language.

Technically, I'm not asking for any changes to anything. I'm happy with the existing syntax. I'm just asking playing devil's advocate :)


> 
> > @pricerc Ok, but if thingThatCanBeDisposed was changed to a Non-IDisposable object, it wouldn't get destroyed when the method exits (i.e. the caller of CreateStuff will still be able to use it), because the GC knows that it is still in use.
> 
> You're confusing IDisposable with garbage collection. In that example, it doesn't matter whether `thingThatCanBeDisposed` is `IDisposable` or not, the garbage collector is _not_ going to collect it when `CreateStuff()` returns because a reference to it still exists inside the `Stuff` instance.
> 

No I'm not. I've implemented it plenty, but I think you're missing my point (which I probably made badly).

I'm just suggesting that a returning method should know whether or not an IDisposable item has a reference that exists outside the method, and so whether or not it said item be disposed. One could argue that if the GC is going to collect something when the method returns (conceptually), then it may as well be Dispose()d; but if it's isn't going to be collected, then don't Dispose() it.

conceptually:
```VB
Public Function CreateStuff() As Stuff
    Dim thingThatCanBeDisposed As IDisposable
    Dim stuff As Stuff

    try
        thingThatCanBeDisposed = New DisposableThing()
        stuff = New Stuff(thingThatCanBeDisposed)

        Return stuff
    finally
        If GC.PlanningToCollect(thingThatCanBeDisposed) then 'not an actual method.
             thingThatCanBeDisposed.Dispose()
        End If
    end try
End Sub
```

**Comment by jdmichel** [comment]
First, I'm still most interested in `Use` as a way to avoid `Using` blocks.

However, I think that @pricerc makes a good point that this feels like something the compiler **may** be able to support, albeit with probable CLR changes.  The GC itself never collects garbage when a method returns. It only collects when an allocation is made that exhausts the gen0 pool (if I remember correctly). So we're really probably talking about adding a feature that can detect when a variable owns a reference object and that variable goes out of scope. If the variable is returned then it's the calling codes job to figure it out, and if the variable is passed to a function then it's that function's job to somehow mark the variable as no longer the owner if it stores a reference. It feels like we're really building back in reference counting to the language, which is how VB used to work prior to .NET and garbage collection.  This feature sounds like a new kind of alternative to GC managed objects. 
At the least this seems like something worthy of a separate proposal.

I still think `Use` could be a worthwhile minor addition, and it's probably why C# added it into 8.0.
